File: practices/practice2.1/src/UserHistory.cpp
#include "../include/UserHistory.hpp"
#include <fstream>

UserHistory::UserHistory(const std::string& filename) : filename(filename) {
    loadHistory();
}

void UserHistory::loadHistory() {
    std::ifstream in(filename);
    std::string name;
    int count;
    while (in >> name >> count) {
        userCount[name] = count;
    }
}

void UserHistory::saveHistory() {
    std::ofstream out(filename);
    for (const auto& [name, count] : userCount) {
        std::println(out, "{} {}", name, count);
    }
}

void UserHistory::resetUser(const std::string& name) {
    userCount.erase(name);
    saveHistory();
}

void UserHistory::resetAllUsers() {
    userCount.clear();
    saveHistory();
}

int UserHistory::incrementFor(const std::string& name) {
    int newCount = ++userCount[name];
    saveHistory();
    return newCount;
}

File: homeworks/homework2.1/src/UserHistory.java
import java.io.*;
import java.util.HashMap;
import java.util.Map;

public class UserHistory {
    private static final String filename = "data/history.txt";
    private static Map<String, Integer> userCount = new HashMap<>();

    public static void loadHistory() throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(" ");
                if (parts.length == 2) {
                    userCount.put(parts[0], Integer.parseInt(parts[1]));
                }
            }
        }
    }

    public static void saveHistory() throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (Map.Entry<String, Integer> entry : userCount.entrySet()) {
                writer.write(entry.getKey() + " " + entry.getValue());
                writer.newLine();
            }
        }
    }

    public static void resetUser(String name) throws IOException {
        userCount.remove(name);
        saveHistory();
    }

    public static void resetAllUsers() throws IOException {
        userCount.clear();
        saveHistory();
    }

    public static void greetUser(String name) throws IOException {
        if (!userCount.containsKey(name)) {
            System.out.println("Welcome, " + name + "!");
        } else {
            System.out.println("Hello again(x" + userCount.get(name) + "), " + name);
        }
        userCount.put(name, userCount.getOrDefault(name, 0) + 1);
        saveHistory();
    }
}
